sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/display/DisplayData.java

```diff
@@ -364,114 +408,202 @@ public String toString() {
+      }
+
+      Item item = Item.create(this, builder.currentPath());
+      Identifier id = item.getIdentifier();
+      checkArgument(!builder.entries.containsKey(id),
+          "Display data key (%s) is not unique within the specified path and namespace: %s%s.",
+          item.getKey(), item.getPath(), item.getNamespace());
+      builder.entries.put(id, item);
+    }
+  }
+
+  /**
+   * Specifies an {@link Item} to register as display data. Each item is identified by a given
+   * path, key, and namespace from the component the display item belongs to.
+   *
+   * <p>{@link Item Items} are registered via {@link DisplayData.Builder#add}
+   * within {@link HasDisplayData#populateDisplayData} implementations.
+   */
+  public static class ItemSpec<T> extends ItemSpecBase<T, ItemSpec<T>> {
+    private ItemSpec(String key, Type type, @Nullable T value) {
+      super(key, type, value);
+    }
+
+    @Override
+    ItemSpec<T> self() {
+      return this;
     }
 
     /**
-     * Set the item {@link Item#getLinkUrl() link url}.
-     *
-     * <p>Specifying a null value will clear the link url if it was previously defined.
-     *
-     * <p>This method does not alter the current instance, but instead returns a new
-     * {@link ItemSpec} with the link url set.
+     * Determine if the given value is equivalent to the raw value representing this spec. Useful
+     * to compare against a specified default value in
+     * {@link Builder#addIfNotDefault}.
      */
-    public ItemSpec<T> withLinkUrl(@Nullable String url) {
-      return toBuilder()
-          .setLinkUrl(url)
-          .build();
+    @Override
+    boolean rawValueEquals(@Nullable T rawValue) {
+      Object otherValue = getType().safeFormat(rawValue).getLongValue();
+      return Objects.equals(getValue(), otherValue);
+    }
+  }
+
+  /**
+   * Specifies a nested sub-component to register as display data.
+   *
+   * @see ItemSpec
+   */
+  public static final class NestedItemSpec extends ItemSpecBase<HasDisplayData, NestedItemSpec> {
+    @Nullable private final HasDisplayData subComponent;
+    private boolean addClassItem = true;
+
+    private NestedItemSpec(String path, @Nullable HasDisplayData subComponent) {
+      super(path, Type.JAVA_CLASS, subComponent == null ? null : subComponent.getClass());
+      this.subComponent = subComponent;
     }
 
     /**
-     * Creates a similar item to the current instance but with the specified value.
-     *
-     * <p>This should only be used internally. It is useful to compare the value of a
-     * {@link DisplayData.Item} to the value derived from a specified input.
+     * Set the class identity to register for the nested subcomponent. To exclude a class item
+     * for the subcomponent, use {@link #withoutClassItem()}.
      */
-    private ItemSpec<T> withValue(T value) {
-      return toBuilder()
-          .setRawValue(value)
-          .build();
+    public NestedItemSpec withClassItem(Class<?> clazz) {
```

Comment:
Naming bikeshed on `.withClassItem(..)` and `.withoutClassItem(..)`.. go!

The idea here: using `DisplayItem.nested(HasDisplayData subcomponent)` will register display data items for (a) the sub-component identity (`subComponent.getClass()`), and (b) all display data held by the subcomponent itself. These overrides allow you to configure the identity that gets registered (a).

Some other ideas:

1. `.withClass(Class<?>)` / `.withoutClass()`: Consistently drop the reference to "item". My worry is that `.withoutClass()` is a confusing name without the context given by `.withClass(Class<?>)`
2. `.withIdentity(Class<?>)` / `.withoutIdentityItem()`: Refer to "identity" rather than "class"